File: ruby/src/grid.rb
require './population.rb'
require './state.rb'

class Grid
    NB_STATES   = 4
    GRID_HEIGHT = 300
    GRID_WIDTH  = 300
  
    attr_accessor :width, :height, :states, :grid_population, :zone
  
    def initialize
      @width  = GRID_WIDTH
      @height = GRID_HEIGHT
      @states = [0] * NB_STATES
      @grid_population = Population.new
      @grid_population.initialize_population
  
      @zone = Array.new(GRID_WIDTH) {
        Array.new(GRID_HEIGHT) { [] }
      }
    end
  
    def initialize_grid(population)
      population.agents.each do |agent|
        x = agent.position.x
        y = agent.position.y
  
        if x < GRID_WIDTH && y < GRID_HEIGHT
          @zone[x][y] << agent.clone
        end
      end
    end
  
    def move_population
      clear_grid
  
      @grid_population.agents.each do |agent|
        agent.move_agent
  
        x = agent.position.x
        y = agent.position.y
  
        if x < @width && y < @height
          @zone[x][y] << agent.clone
        end
      end
    end
  
    def clear_grid
      @zone.each { |column| column.each(&:clear) }
    end
  
    def clear_grid_states
      @states = [0] * NB_STATES
    end
  
    def display_grid
      @zone.each do |column|
        column.each do |cell|
          cell.each { |agent| print agent.get_char_state }
          print "0" if cell.empty?
        end
        puts
      end
    end
  
    def simulate
      clear_grid_states
      move_population
  
      to_infect = []
      to_check = []
  
      @grid_population.agents.each_with_index do |agent, index|
        if agent.state == State::S
          x = agent.position.x
          y = agent.position.y
          probability = probability_to_get_infected_at(x, y)
          to_infect << [index, probability]
        else
          to_check << index
        end
      end
  
      to_infect.each do |index, probability|
        agent = @grid_population.agents[index]
        agent.infection(probability)
      end
  
      to_check.each do |index|
        agent = @grid_population.agents[index]
        agent.check_state
      end
  
      @grid_population.agents.each do |agent|
        agent.increase_time_in_state
        case agent.state
        when State::S then @states[0] += 1
        when State::E then @states[1] += 1
        when State::I then @states[2] += 1
        when State::R then @states[3] += 1
        end
      end
    end
  
    def probability_to_get_infected_at(x, y)
      infectious_neighbors = 0
  
      (-1..1).each do |dx|
        (-1..1).each do |dy|
          neighbor_x = (x + dx + @width) % @width
          neighbor_y = (y + dy + @height) % @height
  
          @zone[neighbor_x][neighbor_y].each do |agent|
            infectious_neighbors += 1 if agent.state == :I
          end
        end
      end
  
      1.0 - Math.exp(-0.5 * infectious_neighbors)
    end

    def get_state(index)
      if index >= 0 && index < NB_STATES
        states[index]
      else
        -1
      end
    end

    def get_width
      @width
    end

    def get_height
      @height
    end

  end  

File: rust/src/grid.rs
use std::collections::LinkedList;
use std::convert::TryInto;
use super::*;

pub const NB_STATES: usize = 4;
pub const GRID_HEIGHT: usize = 300;
pub const GRID_WIDTH: usize = 300;

pub struct Grid
{
    width: usize,
    height: usize,
    states: [usize; NB_STATES],
    grid_population: Population,
    zone: Vec<Vec<LinkedList<Agent>>>
}

impl Grid 
{
    pub fn new() -> Self {
        let states = [0; NB_STATES]; 
        let mut grid_population = Population {
            agents: Vec::with_capacity(NUM_AGENTS),
            duration_time: 0, 
        };

        grid_population.initialize_population();

        let zone = (0..GRID_WIDTH).map(|_| {
            (0..GRID_HEIGHT).map(|_| LinkedList::new()).collect::<Vec<LinkedList<Agent>>>()
        }).collect::<Vec<Vec<LinkedList<Agent>>>>();

        Grid {
            width: GRID_WIDTH,
            height: GRID_HEIGHT,
            states,
            grid_population,
            zone,
        }
    }

    pub fn initialize_grid(&mut self, population: &mut Population) {
        for agent in population.get_agents() {
            let x = agent.get_position().x as usize;
            let y = agent.get_position().y as usize;

            if x < GRID_WIDTH.try_into().unwrap() && y < GRID_HEIGHT.try_into().unwrap() {
                self.zone[x][y].push_back(agent.clone());
            }
        }
    }

    pub fn move_population(&mut self) {
        self.clear_grid();

        let agents_count = self.grid_population.agents.len();
        for i in 0..agents_count {
            if let Some(agent) = self.grid_population.get_mut_agent(i) {
                agent.move_agent(); 

                let x = agent.get_position().x as usize;
                let y = agent.get_position().y as usize;

                if x < self.width && y < self.height {
                    self.zone[x][y].push_back(agent.clone());
                }
            }
        }
    }

    pub fn clear_grid(&mut self) {
        for x in 0..self.width {
            for y in 0..self.height {
                self.zone[x][y].clear();
            }
        }
    }

    pub fn clear_grid_states(&mut self) {
        for i in 0..NB_STATES {
            self.states[i] = 0;
        }
    }

    pub fn display_grid(&self) {
        for x in 0..self.width {
            for y in 0..self.height {
                let list_agent = &self.zone[x][y];
                for e in list_agent {
                    print!("{}", e.get_char_state());
                }
                if list_agent.is_empty() {
                    print!("0");
                }
            }
            println!();
        }
    }

    pub fn simulate(&mut self) {
        self.clear_grid_states();
        self.move_population();

        let mut to_infect = Vec::new();
        let mut to_check  = Vec::new();

        for (i, agent) in self.grid_population.get_agents().iter().enumerate() {
            let state = agent.get_num_state();
            if state == State::S {
                let x = agent.get_position().x as usize;
                let y = agent.get_position().y as usize;
                let probability = self.probability_to_get_infected_at(x, y);
                to_infect.push((i, probability));
            }
            else
            {
               to_check.push(i);
            }
        }

        for (index, probability) in to_infect {
            if let Some(agent) = self.grid_population.get_mut_agent(index) {
                agent.infection(probability);
            }
        }

        for index in to_check
        {
            if let Some(agent) = self.grid_population.get_mut_agent(index) {
                agent.check_state();
            }
        }

        let agents_count = self.grid_population.get_agents().len();
        for i in 0..agents_count {
            if let Some(agent) = self.grid_population.get_mut_agent(i) {
                agent.increase_time_in_state();
                match agent.get_num_state() {
                    State::S => self.states[0] += 1,
                    State::E => self.states[1] += 1,
                    State::I => self.states[2] += 1,
                    State::R => self.states[3] += 1,
                }
            }
        }
    }

    pub fn probability_to_get_infected_at(&self, x: usize, y: usize) -> f64 {
        let mut infectious_neighbors = 0;

        for dx in -1..=1 {
            for dy in -1..=1 {
                let neighbor_x = (x as i32 + dx + self.width as i32) as usize % self.width;
                let neighbor_y = (y as i32 + dy + self.height as i32) as usize % self.height;

                if let Some(agents) = self.zone.get(neighbor_x).and_then(|row| row.get(neighbor_y)) {
                    for agent in agents {
                        if agent as *const _ != self.zone[x][y].front().unwrap() as *const _ && agent.get_num_state() == State::I {
                            infectious_neighbors += 1;
                        }
                    }
                }
            }
        }

        let probability = 1.0 - (-0.5 * infectious_neighbors as f64).exp();
        probability
    }

    pub fn get_width(&self) -> usize {
        self.width
    }

    pub fn get_height(&self) -> usize {
        self.height
    }

    pub fn get_state(&self, index: usize) -> Option<i32> {
        if index < NB_STATES {
            Some(self.states[index].try_into().unwrap())
        } else {
            None
        }
    }
}